src/main/java/uk/gov/hmcts/darts/arm/service/impl/UnstructuredToArmProcessorImpl.java

```diff
@@ -0,0 +1,206 @@
+            failedArmExternalObjectDirectoryEntities.stream()).toList();
+
+        for (var currentExternalObjectDirectoryEntity : allPendingUnstructuredToArmEntities) {
+
+            ExternalObjectDirectoryEntity unstructuredExternalObjectDirectoryEntity;
+            ExternalObjectDirectoryEntity armExternalObjectDirectoryEntity;
+
+            if (currentExternalObjectDirectoryEntity.getExternalLocationType().getId().equals(armLocation.getId())) {
+                armExternalObjectDirectoryEntity = currentExternalObjectDirectoryEntity;
+
+                var matchingEntity = getUnstructuredExternalObjectDirectoryEntity(armExternalObjectDirectoryEntity);
+                if (matchingEntity.isPresent()) {
+                    unstructuredExternalObjectDirectoryEntity = matchingEntity.get();
+                } else {
+                    updateTransferAttempts(armExternalObjectDirectoryEntity);
+                    externalObjectDirectoryRepository.saveAndFlush(armExternalObjectDirectoryEntity);
+                    log.error("Unable to get external object");
+                    continue;
+                }
+            } else {
+                unstructuredExternalObjectDirectoryEntity = currentExternalObjectDirectoryEntity;
+                armExternalObjectDirectoryEntity = createArmExternalObjectDirectoryEntity(currentExternalObjectDirectoryEntity);
+            }
+
+            armExternalObjectDirectoryEntity.setStatus(objectDirectoryStatusRepository.getReferenceById(ARM_INGESTION.getId()));
+            externalObjectDirectoryRepository.saveAndFlush(armExternalObjectDirectoryEntity);
+
+            saveToArm(unstructuredExternalObjectDirectoryEntity, armExternalObjectDirectoryEntity);
+
+            externalObjectDirectoryRepository.saveAndFlush(armExternalObjectDirectoryEntity);
+        }
+    }
+
+    private static List<ObjectRecordStatusEntity> getArmStatuses(ObjectRecordStatusEntity storedStatus,
+                                                                 ObjectRecordStatusEntity failedArmStatus,
+                                                                 ObjectRecordStatusEntity armIngestionStatus) {
+        List<ObjectRecordStatusEntity> armStatuses = new ArrayList<>();
+        armStatuses.add(storedStatus);
+        armStatuses.add(failedArmStatus);
+        armStatuses.add(armIngestionStatus);
+        return armStatuses;
+    }
+
+    private void saveToArm(ExternalObjectDirectoryEntity unstructuredExternalObjectDirectoryEntity,
+                           ExternalObjectDirectoryEntity armExternalObjectDirectoryEntity) {
+        try {
+            String filename = generateFilename(armExternalObjectDirectoryEntity);
+            BinaryData inboundFile = dataManagementApi
+                .getBlobDataFromUnstructuredContainer(unstructuredExternalObjectDirectoryEntity.getExternalLocation());
+
+            armDataManagementApi.saveBlobDataToArm(filename, inboundFile);
```

Comment:
Is this just copying from one Blob container to another, by downloading and then uploading?

Would the following be more appropriate: https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blob-copy-java. From a quick reading it looks like it does the copy server-side which would be a more robust and performant approach?

If that's not appropriate, can you consider using the `blobClient` methods that use streams per https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1694706043#BestPractices(Backend)-2.1Favourjava.ioStreamsOverLoadingLargeDataintoMemory?

This will help avoid reading the full binary data into memory, which might be considerable given the size of an audio file.

Reply:
Its copying from our storage container to an arm storage container. I am in the process of trying to write a functional test for it to see if this does work. This functionality is going to be changed as it needs to use a different way of connecting so it might be best that the optimisation is moved to the ticket when its reworked so that other tickets dependent on this at the very minimum is working. We might have to do something like this: https://stackoverflow.com/questions/61117069/how-to-move-azure-blob-from-one-storage-container-to-another-using-java-rest?